dask/array/core.py

```diff
@@ -234,37 +239,58 @@ def getem(
     lock=False,
     asarray=True,
     dtype=None,
+    inline_array=True,
 ):
     """Dask getting various chunks from an array-like
 
-    >>> getem('X', chunks=(2, 3), shape=(4, 6))  # doctest: +SKIP
+    >>> graph_from_arraylike('X', chunks=(2, 3), shape=(4, 6))  # doctest: +SKIP
     {('X', 0, 0): (getter, 'X', (slice(0, 2), slice(0, 3))),
      ('X', 1, 0): (getter, 'X', (slice(2, 4), slice(0, 3))),
      ('X', 1, 1): (getter, 'X', (slice(2, 4), slice(3, 6))),
      ('X', 0, 1): (getter, 'X', (slice(0, 2), slice(3, 6)))}
 
-    >>> getem('X', chunks=((2, 2), (3, 3)))  # doctest: +SKIP
+    >>> graph_from_arraylike('X', chunks=((2, 2), (3, 3)))  # doctest: +SKIP
     {('X', 0, 0): (getter, 'X', (slice(0, 2), slice(0, 3))),
      ('X', 1, 0): (getter, 'X', (slice(2, 4), slice(0, 3))),
      ('X', 1, 1): (getter, 'X', (slice(2, 4), slice(3, 6))),
      ('X', 0, 1): (getter, 'X', (slice(0, 2), slice(3, 6)))}
     """
     out_name = out_name or arr
     chunks = normalize_chunks(chunks, shape, dtype=dtype)
-    keys = product([out_name], *(range(len(bds)) for bds in chunks))
-    slices = slices_from_chunks(chunks)
 
-    if (
-        has_keyword(getitem, "asarray")
-        and has_keyword(getitem, "lock")
-        and (not asarray or lock)
-    ):
-        values = [(getitem, arr, x, asarray, lock) for x in slices]
+    # If we are inlining the array-like, we can make it an embarassingly parallel
+    # blockwise operation.
+    if inline_array:
+        if (
+            has_keyword(getitem, "asarray")
+            and has_keyword(getitem, "lock")
+            and (not asarray or lock)
+        ):
+            getter = partial(getitem, arr, asarray=asarray, lock=lock)
+        else:
+            # Common case, drop extra parameters
+            getter = partial(getitem, arr)
+
+        return BlockwiseCreateArray(
+            out_name, partial(_get_from_block_info, getter), shape, chunks
+        )
     else:
```

Comment:
I don't love this duplicated logic, but did this in the interests of keeping the option of *not* inlining arrays (at least for the time being). So there are two branches, one that creates an embarrassingly parallel `Blockwise` layer, and another that creates the coupled non-inlined-graph.

Reply:
Handling inlining is pretty straightforward when you make a few small changes:
1. Have `graph_from_arraylike` return a HLG instead of a Layer (since you need 2 layers for the `inline=False` case
2. Go with https://github.com/dask/dask/pull/7417/files#r770188473, and split up `CreateArrayDeps` into multiple little blockwise-dep helpers. One of those helpers would be `Slices`, which produces the `slice` for the current position in the array (essentially replacing your current `_get_from_block_info`).
3. Use `blockwise` directly instead of `BlockwiseCreateArray`. This also spares you needing to embed `arr` in the function via `partial`. The `arr` can be given as a literal blockwise argument, because it'll get [embedded in a SubgraphCallable](https://github.com/dask/dask/blob/27a7928fb750d604dfbad0b3f4361aa64a2c97b4/dask/blockwise.py#L513-L519) and [serialized](https://github.com/dask/dask/blob/27a7928fb750d604dfbad0b3f4361aa64a2c97b4/dask/blockwise.py#L552-L556) as part of `__dask_distributed_pack__`.

```python
if (
    has_keyword(getitem, "asarray")
    and has_keyword(getitem, "lock")
    and (not asarray or lock)
):
    getter = partial(getitem, asarray=asarray, lock=lock)
else:
    # Common case, drop extra parameters
    getter = getitem

layers = {}
deps = {}

slices = Slices(chunks)  # see https://github.com/dask/dask/blob/a154dae7f9893965a95ecd3dde6ec0fa6868764c/dask/blockwise.py#L259-L276
out_indices = tuple(range(len(chunks)))

if inline_array:
    # Tasks become: (getter, arr, slices[idx]) for idx in new_array_chunks
    layers[out_name] = blockwise(getter, out_name, out_indices, arr, None, slices, out_indices)
    deps[out_name] = set()
else:
    base_name = f"{out_name}-array"
    layers[base_name] = MaterializedLayer({base_name: arr})
    deps[base_name] = set()

    # Tasks become: (getter, arr_key, slices[idx]) for idx in new_array_chunks
    layers[out_name] = blockwise(getter, out_name, out_indices, base_name, None, slices, out_indices)
    deps[out_name] = {base_name}

return HighLevelGraph(layers, deps)
```